mlir/lib/Conversion/RockToGPU/RockToGPU.cpp

```diff
@@ -330,6 +323,76 @@ void LowerRockOpsToGPUPass::runOnOperation() {
       signalPassFailure();
   });
 
+  // Post processing the created gpuFuncs
+  op.walk([](gpu::GPUFuncOp gpuFunc) {
+    // Calculate lds usage
+    int64_t ldsUsage = 0;
+    for (const auto &en : llvm::enumerate(gpuFunc.getWorkgroupAttributions())) {
+      BlockArgument ldsBuf = en.value();
+      ShapedType ldsBufType = ldsBuf.getType().cast<ShapedType>();
+      Type elemType = ldsBufType.getElementType();
+      int64_t numElems = ldsBufType.getNumElements();
+      int64_t sizeBytes = (elemType.getIntOrFloatBitWidth() / 8) * numElems;
+      ldsUsage += sizeBytes;
+    }
+    OpBuilder b(gpuFunc.getContext());
+    // The following attribute is set to be used by check-residency pass
+    // later on.
+    if (ldsUsage != 0) {
+      gpuFunc->setAttr("rock.shared_buffer_size",
+                       b.getI32IntegerAttr(ldsUsage));
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Attempting to set wavesPerEU...\n");
+    if (!gpuFunc->hasAttrOfType<IntegerAttr>("block_size")) {
+      LLVM_DEBUG(llvm::dbgs() << "blockSize not found in gpuFunc.\n");
+      return;
+    }
+    int64_t blockSize =
+        gpuFunc->getAttrOfType<IntegerAttr>("block_size").getInt();
+    if (!gpuFunc->hasAttrOfType<IntegerAttr>("grid_size")) {
+      LLVM_DEBUG(llvm::dbgs() << "gridSize not found in gpuFunc.\n");
+      return;
+    }
+    int64_t gridSize =
+        gpuFunc->getAttrOfType<IntegerAttr>("grid_size").getInt();
+    StringAttr arch = rock::getArch(gpuFunc);
+    if (arch) {
+      rock::AmdArchInfo archInfo = rock::lookupArchInfo(arch);
+      int64_t wavesPerBlock =
+          blockSize / (archInfo.numEUPerCU * archInfo.waveSize);
+      int64_t wgsPerCU = gridSize / archInfo.minNumCU;
+      LLVM_DEBUG(llvm::dbgs() << "wavesPerBlock:" << wavesPerBlock << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "  blockSize:" << blockSize << "\n");
+      LLVM_DEBUG(llvm::dbgs()
+                 << "  numEUPerCU:" << archInfo.numEUPerCU << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "  waveSize:" << archInfo.waveSize << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "wgsPerCU:" << wgsPerCU << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "  gridSize:" << gridSize << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "  numCUs:" << archInfo.minNumCU << "\n");
+      LLVM_DEBUG(llvm::dbgs()
+                 << "maxSharedMemPerWG:" << archInfo.maxSharedMemPerWG << "\n");
+      LLVM_DEBUG(llvm::dbgs() << "ldsUsage:" << ldsUsage << "\n");
+      bool hasMoreThan2WavesPerBlock = wavesPerBlock >= 2;
+      bool hasMoreThan2WGsPerCU = wgsPerCU >= 2;
+      bool isUsingLessThanHalfLds = true;
+      if (ldsUsage > 0) {
+        isUsingLessThanHalfLds = archInfo.maxSharedMemPerWG / ldsUsage >= 2;
```

Comment:
This also doesn't account for maxSharedMemPerCU which is different than maxSharedMemPerWG on Navi. A WG could use all of maxSharedMemPerWG and still have 2 WGs per CU.

Reply:
I see. so I should really be using maxSharedMemPerCU here...

Reply:
done